File: tests/timer-logic.test.js
const T = require("../lib/timer-logic.js");

// ---- formatMMSS ----
test("formatMMSS pads seconds", () => {
  assert.eq(T.formatMMSS(0), "0:00");
  assert.eq(T.formatMMSS(5), "0:05");
  assert.eq(T.formatMMSS(65), "1:05");
  assert.eq(T.formatMMSS(300), "5:00");
});

test("formatMMSS clamps negatives to zero", () => {
  assert.eq(T.formatMMSS(-10), "0:00");
});

// ---- clampWorkMinutes ----
test("clampWorkMinutes enforces bounds", () => {
  assert.eq(T.clampWorkMinutes(25), 25);
  assert.eq(T.clampWorkMinutes(0), 1);
  assert.eq(T.clampWorkMinutes(-7), 1);
  assert.eq(T.clampWorkMinutes(9999), 120);
  assert.eq(T.clampWorkMinutes("45"), 45);
});

test("clampWorkMinutes falls back on garbage", () => {
  assert.eq(T.clampWorkMinutes("nope"), T.DEFAULT_WORK_MINUTES);
  assert.eq(T.clampWorkMinutes(NaN), T.DEFAULT_WORK_MINUTES);
});

// ---- weightStageForHours ----
test("weight stages match the spec", () => {
  assert.eq(T.weightStageForHours(0).stage, 0);
  assert.eq(T.weightStageForHours(0.99).stage, 0);
  assert.eq(T.weightStageForHours(1).stage, 1);
  assert.eq(T.weightStageForHours(2).stage, 2);
  // hour 3 -> takes ~half the screen
  assert.eq(T.weightStageForHours(3).stage, 3);
  assert.truthy(T.weightStageForHours(3).sizeFactor >= 0.5);
  // hour 5 -> can barely see the code
  assert.eq(T.weightStageForHours(5).stage, 4);
  assert.truthy(T.weightStageForHours(5).sizeFactor >= 0.8);
});

test("weight is monotonically non-decreasing in hours", () => {
  let prev = 0;
  for (const h of [0, 0.5, 1, 1.5, 2, 2.5, 3, 4, 5, 8]) {
    const f = T.weightStageForHours(h).sizeFactor;
    assert.truthy(f >= prev, `regressed at h=${h}`);
    prev = f;
  }
});

// ---- stretchFactor ----
test("stretchFactor opens from 1.0 at 5:00 to 1.6 by 4:47", () => {
  const breakSec = 300;
  assert.approx(T.stretchFactor(300, breakSec), 1.0, 1e-6);
  assert.approx(T.stretchFactor(287, breakSec), 1.6, 1e-6);
  // halfway through the 13s window
  assert.approx(T.stretchFactor(294, breakSec), 1.0 + (6 / 13) * 0.6, 1e-6);
});

test("stretchFactor stays at 1.6 after 4:47 (the cat is comfortable)", () => {
  assert.approx(T.stretchFactor(200, 300), 1.6, 1e-6);
  assert.approx(T.stretchFactor(0, 300), 1.6, 1e-6);
});

// ---- remainingSeconds ----
test("remainingSeconds computes ceil of phase delta", () => {
  const now = 1_000_000;
  const state = { phase: "break", phaseEndsAt: now + 4500 };
  assert.eq(T.remainingSeconds(state, now), 5);
});

test("remainingSeconds floors at 0 once phase has elapsed", () => {
  const now = 1_000_000;
  const state = { phase: "break", phaseEndsAt: now - 1000 };
  assert.eq(T.remainingSeconds(state, now), 0);
});

test("remainingSeconds returns 0 when idle", () => {
  assert.eq(T.remainingSeconds({ phase: "idle" }, 1), 0);
});

// ---- continuousWorkHours ----
test("continuousWorkHours counts from streak start", () => {
  const start = 0;
  const oneHour = 60 * 60 * 1000;
  assert.approx(T.continuousWorkHours(start, oneHour), 1, 1e-6);
  assert.approx(T.continuousWorkHours(start, 3 * oneHour), 3, 1e-6);
});

test("continuousWorkHours is 0 when no streak is set", () => {
  assert.eq(T.continuousWorkHours(null, Date.now()), 0);
});

// ---- advancePhase: idle -> work -> break -> work ----
test("advancePhase from idle starts a work block", () => {
  const now = 1_000_000;
  const next = T.advancePhase(
    { phase: "idle", phaseEndsAt: 0, workStreakStartedAt: null },
    { workMinutes: 25 },
    now
  );
  assert.eq(next.phase, "work");
  assert.eq(next.phaseEndsAt, now + 25 * 60 * 1000);
  assert.eq(next.workStreakStartedAt, now);
});

test("advancePhase from work transitions to break and preserves streak", () => {
  const streakStart = 1_000_000;
  const now = streakStart + 25 * 60 * 1000;
  const next = T.advancePhase(
    { phase: "work", phaseEndsAt: now, workStreakStartedAt: streakStart },
    { workMinutes: 25 },
    now
  );
  assert.eq(next.phase, "break");
  assert.eq(next.phaseEndsAt, now + 5 * 60 * 1000);
  assert.eq(next.workStreakStartedAt, streakStart, "streak must persist into break");
  assert.eq(next.breakStartedAt, now);
});

test("advancePhase from break resets the streak — the only way to shrink the cat", () => {
  const streakStart = 0;
  const breakEnd = streakStart + 30 * 60 * 1000;
  const next = T.advancePhase(
    { phase: "break", phaseEndsAt: breakEnd, workStreakStartedAt: streakStart },
    { workMinutes: 25 },
    breakEnd
  );
  assert.eq(next.phase, "work");
  assert.eq(next.workStreakStartedAt, breakEnd, "streak must reset after a completed break");
});

test("advancePhase respects custom work minutes from settings", () => {
  const now = 0;
  const next = T.advancePhase({ phase: "idle" }, { workMinutes: 50 }, now);
  assert.eq(next.phaseEndsAt, 50 * 60 * 1000);
});

test("advancePhase clamps absurd workMinutes (defends against popup tampering)", () => {
  const now = 0;
  const tooBig = T.advancePhase({ phase: "idle" }, { workMinutes: 100000 }, now);
  assert.eq(tooBig.phaseEndsAt, T.MAX_WORK_MINUTES * 60 * 1000);
  const tooSmall = T.advancePhase({ phase: "idle" }, { workMinutes: 0 }, now);
  assert.eq(tooSmall.phaseEndsAt, T.MIN_WORK_MINUTES * 60 * 1000);
});

// ---- Scenario: end-to-end weight gain across a 5h streak ----
test("scenario: 5 hours of unbroken work lands in the eclipse stage", () => {
  const t0 = 0;
  let state = T.advancePhase({ phase: "idle" }, { workMinutes: 25 }, t0);
  // Pretend the user keeps cancelling breaks (well, they can't — but this
  // simulates the streak logic if the break never *completes*).
  const fiveHours = 5 * 60 * 60 * 1000;
  const hours = T.continuousWorkHours(state.workStreakStartedAt, t0 + fiveHours);
  const weight = T.weightStageForHours(hours);
  assert.eq(weight.stage, 4);
  assert.eq(weight.label, "eclipse");
});

// ---- Scenario: timer reaches 0 exactly at break end ----
test("scenario: countdown hits 0:00 exactly at phaseEndsAt", () => {
  const phaseEndsAt = 1_700_000_000_000;
  const state = { phase: "break", phaseEndsAt };
  assert.eq(T.formatMMSS(T.remainingSeconds(state, phaseEndsAt)), "0:00");
  assert.eq(T.formatMMSS(T.remainingSeconds(state, phaseEndsAt - 1000)), "0:01");
});

// ---- Scenario: the only way to shrink the cat is to take the break ----
test("scenario: break completion is the only path to reset weight", () => {
  const t0 = 0;
  // Build up a 4-hour streak.
  const streakStart = t0;
  const fourHours = 4 * 60 * 60 * 1000;
  const beforeBreak = T.weightStageForHours(
    T.continuousWorkHours(streakStart, t0 + fourHours)
  );
  assert.eq(beforeBreak.stage, 3);

  // Break completes -> streak resets -> weight returns to stage 0.
  const afterBreak = T.advancePhase(
    {
      phase: "break",
      phaseEndsAt: t0 + fourHours + 5 * 60 * 1000,
      workStreakStartedAt: streakStart,
    },
    { workMinutes: 25 },
    t0 + fourHours + 5 * 60 * 1000
  );
  const newWeight = T.weightStageForHours(
    T.continuousWorkHours(afterBreak.workStreakStartedAt, afterBreak.workStreakStartedAt)
  );
  assert.eq(newWeight.stage, 0);
});

File: android-app/app/src/test/kotlin/com/fatorangecat/core/TimerLogicTest.kt
package com.fatorangecat.core

import com.fatorangecat.core.TimerLogic.Phase
import com.fatorangecat.core.TimerLogic.Settings
import com.fatorangecat.core.TimerLogic.TimerState
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

private const val EPS = 1e-6

private fun approx(a: Double, b: Double, eps: Double = EPS, msg: String = "") {
    assertTrue("$msg |$a - $b| > $eps", kotlin.math.abs(a - b) <= eps)
}

class TimerLogicTest {

    // ---- formatMMSS ----

    @Test
    fun formatMMSS_padsSeconds() {
        assertEquals("0:00", TimerLogic.formatMMSS(0))
        assertEquals("0:05", TimerLogic.formatMMSS(5))
        assertEquals("1:05", TimerLogic.formatMMSS(65))
        assertEquals("5:00", TimerLogic.formatMMSS(300))
    }

    @Test
    fun formatMMSS_clampsNegativesToZero() {
        assertEquals("0:00", TimerLogic.formatMMSS(-10))
    }

    // ---- clampWorkMinutes ----

    @Test
    fun clampWorkMinutes_enforcesBounds() {
        assertEquals(25, TimerLogic.clampWorkMinutes(25))
        assertEquals(1, TimerLogic.clampWorkMinutes(0))
        assertEquals(1, TimerLogic.clampWorkMinutes(-7))
        assertEquals(120, TimerLogic.clampWorkMinutes(9999))
        assertEquals(45, TimerLogic.clampWorkMinutes("45"))
    }

    @Test
    fun clampWorkMinutes_fallsBackOnGarbage() {
        assertEquals(TimerLogic.DEFAULT_WORK_MINUTES, TimerLogic.clampWorkMinutes("nope"))
        assertEquals(TimerLogic.DEFAULT_WORK_MINUTES, TimerLogic.clampWorkMinutes(Double.NaN))
    }

    // ---- weightStageForHours ----

    @Test
    fun weightStages_matchTheSpec() {
        assertEquals(0, TimerLogic.weightStageForHours(0.0).stage)
        assertEquals(0, TimerLogic.weightStageForHours(0.99).stage)
        assertEquals(1, TimerLogic.weightStageForHours(1.0).stage)
        assertEquals(2, TimerLogic.weightStageForHours(2.0).stage)
        // hour 3 -> takes ~half the screen
        assertEquals(3, TimerLogic.weightStageForHours(3.0).stage)
        assertTrue(TimerLogic.weightStageForHours(3.0).sizeFactor >= 0.5)
        // hour 5 -> can barely see the code
        assertEquals(4, TimerLogic.weightStageForHours(5.0).stage)
        assertTrue(TimerLogic.weightStageForHours(5.0).sizeFactor >= 0.8)
    }

    @Test
    fun weightStages_monotonicallyNonDecreasing() {
        var prev = 0.0
        for (h in listOf(0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 4.0, 5.0, 8.0)) {
            val f = TimerLogic.weightStageForHours(h).sizeFactor
            assertTrue("regressed at h=$h", f >= prev)
            prev = f
        }
    }

    // ---- stretchFactor ----

    @Test
    fun stretchFactor_opensFrom1to1_6Across13Seconds() {
        val breakSec = 300
        approx(TimerLogic.stretchFactor(300, breakSec), 1.0)
        approx(TimerLogic.stretchFactor(287, breakSec), 1.6)
        // halfway through the 13s window
        approx(TimerLogic.stretchFactor(294, breakSec), 1.0 + (6.0 / 13.0) * 0.6)
    }

    @Test
    fun stretchFactor_staysAt1_6AfterTheCatIsComfortable() {
        approx(TimerLogic.stretchFactor(200, 300), 1.6)
        approx(TimerLogic.stretchFactor(0, 300), 1.6)
    }

    // ---- remainingSeconds ----

    @Test
    fun remainingSeconds_computesCeilOfPhaseDelta() {
        val now = 1_000_000L
        val state = TimerState(phase = Phase.BREAK, phaseEndsAt = now + 4500L)
        assertEquals(5, TimerLogic.remainingSeconds(state, now))
    }

    @Test
    fun remainingSeconds_floorsAtZeroOnceElapsed() {
        val now = 1_000_000L
        val state = TimerState(phase = Phase.BREAK, phaseEndsAt = now - 1000L)
        assertEquals(0, TimerLogic.remainingSeconds(state, now))
    }

    @Test
    fun remainingSeconds_returnsZeroWhenIdle() {
        assertEquals(0, TimerLogic.remainingSeconds(TimerState(phase = Phase.IDLE), 1L))
    }

    // ---- continuousWorkHours ----

    @Test
    fun continuousWorkHours_countsFromStreakStart() {
        val start = 0L
        val oneHour = 60L * 60L * 1000L
        approx(TimerLogic.continuousWorkHours(start, oneHour), 1.0)
        approx(TimerLogic.continuousWorkHours(start, 3L * oneHour), 3.0)
    }

    @Test
    fun continuousWorkHours_isZeroWhenNoStreak() {
        assertEquals(0.0, TimerLogic.continuousWorkHours(null, System.currentTimeMillis()), 0.0)
    }

    // ---- advancePhase ----

    @Test
    fun advancePhase_fromIdleStartsAWorkBlock() {
        val now = 1_000_000L
        val next = TimerLogic.advancePhase(
            TimerState(phase = Phase.IDLE, phaseEndsAt = 0L, workStreakStartedAt = null),
            Settings(workMinutes = 25),
            now,
        )
        assertEquals(Phase.WORK, next.phase)
        assertEquals(now + 25L * 60L * 1000L, next.phaseEndsAt)
        assertEquals(now, next.workStreakStartedAt)
    }

    @Test
    fun advancePhase_fromWorkTransitionsToBreakAndPreservesStreak() {
        val streakStart = 1_000_000L
        val now = streakStart + 25L * 60L * 1000L
        val next = TimerLogic.advancePhase(
            TimerState(phase = Phase.WORK, phaseEndsAt = now, workStreakStartedAt = streakStart),
            Settings(workMinutes = 25),
            now,
        )
        assertEquals(Phase.BREAK, next.phase)
        assertEquals(now + 5L * 60L * 1000L, next.phaseEndsAt)
        assertEquals("streak must persist into break", streakStart, next.workStreakStartedAt)
        assertEquals(now, next.breakStartedAt)
    }

    @Test
    fun advancePhase_fromBreakResetsStreak_theOnlyWayToShrinkTheCat() {
        val streakStart = 0L
        val breakEnd = streakStart + 30L * 60L * 1000L
        val next = TimerLogic.advancePhase(
            TimerState(phase = Phase.BREAK, phaseEndsAt = breakEnd, workStreakStartedAt = streakStart),
            Settings(workMinutes = 25),
            breakEnd,
        )
        assertEquals(Phase.WORK, next.phase)
        assertEquals(
            "streak must reset after a completed break",
            breakEnd,
            next.workStreakStartedAt,
        )
    }

    @Test
    fun advancePhase_respectsCustomWorkMinutes() {
        val now = 0L
        val next = TimerLogic.advancePhase(
            TimerState(phase = Phase.IDLE),
            Settings(workMinutes = 50),
            now,
        )
        assertEquals(50L * 60L * 1000L, next.phaseEndsAt)
    }

    @Test
    fun advancePhase_clampsAbsurdWorkMinutes() {
        val now = 0L
        val tooBig = TimerLogic.advancePhase(
            TimerState(phase = Phase.IDLE),
            Settings(workMinutes = 100_000),
            now,
        )
        assertEquals(TimerLogic.MAX_WORK_MINUTES * 60L * 1000L, tooBig.phaseEndsAt)
        val tooSmall = TimerLogic.advancePhase(
            TimerState(phase = Phase.IDLE),
            Settings(workMinutes = 0),
            now,
        )
        assertEquals(TimerLogic.MIN_WORK_MINUTES * 60L * 1000L, tooSmall.phaseEndsAt)
    }

    // ---- shouldWarn ----

    @Test
    fun shouldWarn_firesOnlyInFinalWarningSecondsOfWork() {
        val now = 1_000_000L
        val justStarted = TimerState(phase = Phase.WORK, phaseEndsAt = now + 25L * 60L * 1000L)
        assertFalse(TimerLogic.shouldWarn(justStarted, now))

        val tightWindow = TimerState(phase = Phase.WORK, phaseEndsAt = now + 30L * 1000L)
        assertTrue(TimerLogic.shouldWarn(tightWindow, now))

        val oneSecondLeft = TimerState(phase = Phase.WORK, phaseEndsAt = now + 999L)
        assertTrue(TimerLogic.shouldWarn(oneSecondLeft, now))

        val justEnded = TimerState(phase = Phase.WORK, phaseEndsAt = now)
        assertFalse(TimerLogic.shouldWarn(justEnded, now))
    }

    @Test
    fun shouldWarn_neverFiresDuringBreakOrIdle() {
        val now = 1_000_000L
        assertFalse(
            TimerLogic.shouldWarn(
                TimerState(phase = Phase.BREAK, phaseEndsAt = now + 15L * 1000L),
                now,
            ),
        )
        assertFalse(TimerLogic.shouldWarn(TimerState(phase = Phase.IDLE, phaseEndsAt = 0L), now))
    }

    @Test
    fun warningSeconds_isExposedAndSane() {
        assertTrue(TimerLogic.WARNING_SECONDS in 1..120)
    }

    // ---- Scenarios (end-to-end behavior pinned to spec) ----

    @Test
    fun scenario_fiveHoursOfUnbrokenWorkLandsInEclipse() {
        val t0 = 0L
        val state = TimerLogic.advancePhase(
            TimerState(phase = Phase.IDLE),
            Settings(workMinutes = 25),
            t0,
        )
        // Pretend the user keeps cancelling breaks (they can't — but this
        // simulates the streak logic if the break never *completes*).
        val fiveHours = 5L * 60L * 60L * 1000L
        val hours = TimerLogic.continuousWorkHours(state.workStreakStartedAt, t0 + fiveHours)
        val weight = TimerLogic.weightStageForHours(hours)
        assertEquals(4, weight.stage)
        assertEquals("eclipse", weight.label)
    }

    @Test
    fun scenario_countdownHitsZeroExactlyAtPhaseEndsAt() {
        val phaseEndsAt = 1_700_000_000_000L
        val state = TimerState(phase = Phase.BREAK, phaseEndsAt = phaseEndsAt)
        assertEquals("0:00", TimerLogic.formatMMSS(TimerLogic.remainingSeconds(state, phaseEndsAt)))
        assertEquals(
            "0:01",
            TimerLogic.formatMMSS(TimerLogic.remainingSeconds(state, phaseEndsAt - 1000L)),
        )
    }

    @Test
    fun scenario_forceBreakStartsBreakAndPreservesStreak() {
        // Mirrors the FORCE_BREAK / "Summon him now" code path in
        // background.js / TimerService.handleForceBreak: pretend we're
        // in WORK and advance.
        val streakStart = 1_000_000L
        val now = streakStart + 7L * 60L * 1000L // 7 minutes into the work block
        val current = TimerState(
            phase = Phase.WORK,
            phaseEndsAt = streakStart + 25L * 60L * 1000L,
            workStreakStartedAt = streakStart,
        )
        val next = TimerLogic.advancePhase(
            current.copy(phase = Phase.WORK),
            Settings(workMinutes = 25),
            now,
        )
        assertEquals(Phase.BREAK, next.phase)
        assertEquals(now + TimerLogic.BREAK_MINUTES * 60L * 1000L, next.phaseEndsAt)
        assertEquals(
            "force-break must NOT reset the streak — only completing the break does",
            streakStart,
            next.workStreakStartedAt,
        )
    }

    @Test
    fun scenario_forceBreakFromIdleInitializesStreakToNow() {
        val now = 5_000_000L
        val next = TimerLogic.advancePhase(
            TimerState(phase = Phase.IDLE).copy(phase = Phase.WORK),
            Settings(workMinutes = 25),
            now,
        )
        assertEquals(Phase.BREAK, next.phase)
        assertEquals(
            "force-break from idle must seed the streak so weight starts at 0",
            now,
            next.workStreakStartedAt,
        )
    }

    @Test
    fun scenario_breakCompletionIsTheOnlyPathToResetWeight() {
        val t0 = 0L
        val streakStart = t0
        val fourHours = 4L * 60L * 60L * 1000L
        val beforeBreak = TimerLogic.weightStageForHours(
            TimerLogic.continuousWorkHours(streakStart, t0 + fourHours),
        )
        assertEquals(3, beforeBreak.stage)

        val afterBreak = TimerLogic.advancePhase(
            TimerState(
                phase = Phase.BREAK,
                phaseEndsAt = t0 + fourHours + 5L * 60L * 1000L,
                workStreakStartedAt = streakStart,
            ),
            Settings(workMinutes = 25),
            t0 + fourHours + 5L * 60L * 1000L,
        )
        val newWeight = TimerLogic.weightStageForHours(
            TimerLogic.continuousWorkHours(
                afterBreak.workStreakStartedAt,
                afterBreak.workStreakStartedAt!!,
            ),
        )
        assertEquals(0, newWeight.stage)
    }
}
